FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean

```diff
@@ -547,7 +548,65 @@ noncomputable def ProdAdicCompletions.baseChangeEquiv :
     L ⊗[K] ProdAdicCompletions A K ≃ₐ[L] ProdAdicCompletions B L :=
   AlgEquiv.ofBijective
     (SemialgHom.baseChange_of_algebraMap (ProdAdicCompletions.baseChange A K L B))
-    sorry -- #239
+    (by
+      --strategy, compose a `K`-linear iso from `L ⊗ ∏ K v` to `∏ L w`
+      -- `prod_equiv` says that `∏ L ⊗ K v ≃ ∏v∣w ∏w L w`
+      let prod_equiv := AlgEquiv.piCongrRight (fun (v: HeightOneSpectrum A)
+        ↦ adicCompletionComapAlgEquiv A K L B v)
+      -- `commute` says `L ⊗ ∏ K v ≃ ∏ L ⊗ K v`
+      let commute := tensorPi_equiv_piTensor K L (adicCompletion K (R := A))
+      -- `restrict` map restricts `prod_equiv` to `K`-linear iso
+      let restrict := (prod_equiv.restrictScalars K).toLinearEquiv
+      -- `equiv_prod` gives a map from `L ⊗ ∏ K v ≃ ∏v∣w ∏w L w`
+      let equiv_prod := commute ≪≫ₗ restrict
+      -- not picking up instance; `inst_alg` should be a `haveI`? Fails on next `haveI`
+      let inst_alg : Algebra K (ProdAdicCompletions B L) := RingHom.toAlgebra <|
+        (algebraMap L (ProdAdicCompletions B L)).comp (algebraMap K L)
+      haveI : IsScalarTower K L (ProdAdicCompletions B L) :=
+        IsScalarTower.of_algebraMap_eq (congrFun rfl)
+      -- `prod_equiv'` gives a map from ∏v∣w ∏w L w ≃ ∏ L w
+      let prod_equiv' : (∀ (v : HeightOneSpectrum A),
+          (∀ w : v.Extension B, w.1.adicCompletion L)) ≃ₐ[L] ProdAdicCompletions B L :=
+        {
+        toFun f w := f (comap A w) ⟨w, rfl⟩
+        invFun g _ w := g w.1
+        left_inv f := by
+          ext v w₁
+          simp only
+          congr
+          · exact w₁.2
+          · rw [w₁.2]
+          · exact proof_irrel_heq rfl w₁.2
```

Comment:
```suggestion
        left_inv f := by
          ext v ⟨w, rfl⟩
          simp only
```
Lean can't see through the `Extension B v` because it's a def (so Lean won't unfold it) and nobody wrote any simp lemmas about it (so automation won't work either). If you just take it apart yourself then normal service is resumed.